src/main/java/org/folio/circulation/domain/notice/schedule/RequestScheduledNoticeService.java

```diff
@@ -0,0 +1,132 @@
+import org.folio.circulation.domain.notice.NoticeConfiguration;
+import org.folio.circulation.domain.notice.NoticeEventType;
+import org.folio.circulation.domain.notice.PatronNoticePolicy;
+import org.folio.circulation.domain.policy.PatronNoticePolicyRepository;
+import org.folio.circulation.support.Clients;
+import org.folio.circulation.support.Result;
+import org.joda.time.DateTime;
+
+public class RequestScheduledNoticeService {
+  public static RequestScheduledNoticeService using(Clients clients) {
+    return new RequestScheduledNoticeService(
+      ScheduledNoticesRepository.using(clients),
+      new PatronNoticePolicyRepository(clients));
+  }
+
+  private final ScheduledNoticesRepository scheduledNoticesRepository;
+  private final PatronNoticePolicyRepository noticePolicyRepository;
+
+  private RequestScheduledNoticeService(
+    ScheduledNoticesRepository scheduledNoticesRepository,
+    PatronNoticePolicyRepository noticePolicyRepository) {
+    this.scheduledNoticesRepository = scheduledNoticesRepository;
+    this.noticePolicyRepository = noticePolicyRepository;
+  }
+
+
+  public Result<RequestAndRelatedRecords> scheduleRequestNotices(RequestAndRelatedRecords relatedRecords) {
+    scheduleRequestNotices(relatedRecords.getRequest());
+    return succeeded(relatedRecords);
+  }
+
+  public Result<RequestAndRelatedRecords> rescheduleRequestNotices(RequestAndRelatedRecords relatedRecords) {
+    Request request = relatedRecords.getRequest();
+    scheduledNoticesRepository.deleteByRequestId(request.getId())
+      .thenAccept(r -> r.next(resp -> scheduleRequestNotices(request)));
+
+    return succeeded(relatedRecords);
+  }
+
+  public Result<CheckInProcessRecords> rescheduleRequestNotices(CheckInProcessRecords records) {
+    Optional.ofNullable(records.getHighestPriorityFulfillableRequest())
+      .ifPresent(this::rescheduleRequestNotices);
+
+    return succeeded(records);
+  }
+
+  private void rescheduleRequestNotices(Request request) {
+    scheduledNoticesRepository.deleteByRequestId(request.getId())
+      .thenAccept(r -> r.next(resp -> scheduleRequestNotices(request)));
+  }
+
+  private Result<PatronNoticePolicy> scheduleRequestNoticesBasedOnPolicy(
+    Request request, PatronNoticePolicy noticePolicy) {
+
+    noticePolicy.getNoticeConfigurations()
+      .stream()
+      .filter(cfg -> requiresNoticeScheduling(cfg, request))
+      .map(cfg -> createRequestScheduledNotice(cfg, request))
+      .forEach(scheduledNoticesRepository::create);
+
+    return succeeded(noticePolicy);
+  }
+
+  private DateTime determineNextRunTime(NoticeConfiguration cfg, Request request) {
+    DateTime expirationTime = getExpirationTime(cfg, request);
+    if (cfg.getTiming() == UPON_AT) {
+      return expirationTime;
+    }
+    return expirationTime.minus(cfg.getTimingPeriod().timePeriod());
+  }
+
+  private DateTime getExpirationTime(NoticeConfiguration cfg, Request request) {
+    switch (cfg.getNoticeEventType()) {
```

Comment:
These repeated checks on request expiration vs. hold expiration suggests to me there is a missing abstraction, maybe and event type? (I don't know the domain well enough to advise)

Reply:
Reworked. Separated the logic into two methods for creating request scheduled notices(request expiration and hold expiration).